clang/utils/TableGen/ClangProgModelBuiltinEmitter.cpp

```diff
@@ -848,6 +922,216 @@ void BuiltinNameEmitter::EmitQualTypeFinder() {
+
+  if (ExpansionIt != ExpandedTypes.end())
+    return;
+  llvm::SmallVectorImpl<std::string> &Expansion = ExpandedTypes[Ty];
+
+  JSONBuiltinInterfaceEmitter::TypeDesc TypeDesc(Ty);
+
+  bool IsCompound = Ty->isSubClassOf("CompoundType");
+  if (IsCompound) {
+    auto *EltRecord = Ty->getValueAsDef("ElementType");
+    ExpandType(EltRecord);
+    Expansion = ExpandedTypes[EltRecord];
+    for (auto &TypeStr : Expansion) {
+      if (TypeDesc.IsPointer) {
+        if (TypeDesc.IsConst)
+          TypeStr += " const";
+        if (TypeDesc.IsVolatile)
+          TypeStr += " volatile";
+        TypeStr += " ";
+        TypeStr += TypeDesc.AddrSpace;
+        TypeStr += " *";
+      }
+    }
+
+    return;
+  }
+
+  if (Ty->isSubClassOf("GenericType")) {
+    assert(!Expansion.size() && "Types already expanded");
+    auto NVec = Ty->getValueAsDef("VectorList")->getValueAsListOfInts("List");
+    auto NTypes = Ty->getValueAsDef("TypeList")->getValueAsListOfDefs("List");
+    for (auto SubT : NTypes) {
+      for (auto VecWidth : NVec) {
+        JSONBuiltinInterfaceEmitter::TypeDesc TypeDesc(SubT);
+        TypeDesc.VecWidth = VecWidth;
+        Expansion.emplace_back(TypeDesc.GetBaseTypeAsStr());
+      }
+    }
+
+    return;
+  }
+
+  Expansion.emplace_back(TypeDesc.GetBaseTypeAsStr());
+}
+
+void JSONBuiltinInterfaceEmitter::ExpandTypes() {
+  for (const auto *T : Records.getAllDerivedDefinitions("Type")) {
+    ExpandType(T);
+  }
+}
+
+void JSONBuiltinInterfaceEmitter::EmitBuiltins() {
+  struct FnDesc {
+    bool IsPure;
+    bool IsConst;
+    bool IsConv;
+    bool IsVariadic;
+    SmallVector<llvm::StringRef, 8> Args;
+  };
+  StringMap<SmallVector<FnDesc, 16>> NameToProtoList;
+
+  // For each function names, gather the list of overloads
+  for (const auto &SLM : SignatureListMap) {
+    for (const auto &Name : SLM.second.Names) {
+      auto &PrototypeList = NameToProtoList[Name];
+
+      for (const auto &Overload : SLM.second.Signatures) {
+        auto Signature = Overload.first->getValueAsListOfDefs("Signature");
+        auto SignatureTypesIt = llvm::map_range(
+            Signature, [this](const Record *Ty) -> llvm::ArrayRef<std::string> {
+              return ExpandedTypes[Ty];
+            });
+        auto SignatureLenIt = llvm::map_range(
+            SignatureTypesIt, [](llvm::ArrayRef<std::string> ExpandedType) {
+              return ExpandedType.size();
+            });
+        assert(SignatureLenIt.begin() != SignatureLenIt.end());
+        // Find out how many overloads this signature holds.
+        size_t NbOverload = 1;
+        for (size_t Len : SignatureLenIt)
+          NbOverload = std::max(NbOverload, Len);
+        // assert the signature is consistent
+        assert(llvm::all_of(SignatureLenIt, [NbOverload](size_t NbTypes) {
+          return NbTypes == NbOverload || NbTypes == 1;
+        }));
+        for (size_t idx = 0; idx < NbOverload; idx++) {
+          FnDesc Proto;
```

Comment:
It seems to me that this type should just have a constructor to get these values out.  Probably the same with the other types you added as well.

Reply:
I moved the code.

> other types you added as well

the other one already have a ctor.